### PR TITLE
Adapt interface for orthogonal and symplectic Lie algebras

### PR DESCRIPTION
**Pull request by lgoettgens**
As discussed with @fingolfin in some slack thread a long time ago, the function `special_orthogonal_lie_algebra` should be able to handle an arbitrary symmetric bilinear form as input. This is now possible; the previous behavior can be achieved by providing `identity_matrix(ZZ, n)` as the form.

Furthermore, I added the same interface for the new `symplectic_lie_algebra` function.

### CODE REVIEW

**Comment by lgoettgens** [comment]
~~This will need to be rebased once https://github.com/oscar-system/Oscar.jl/pull/3387 is merged to avoid semantical conflicts.~~

**Comment by lgoettgens** [comment]
Bump @fingolfin

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl [hunk 1]
```diff
@@ -317,37 +321,191 @@ function special_linear_lie_algebra(R::Field, n::Int)
...
+
+Return the symplectic Lie algebra $\mathfrak{sp}_n(R)$.
+
+Given a non-degenerate skew-symmetric bilinear form $f$ via its Gram matrix `gram`,
+$\mathfrak{sp}_n(R)$ is the Lie algebra of all $n \times n$ matrices $x$ over the field `R`
+such that $f(xv, w) = -f(v, xw)$ for all $v, w \in R^n$.
+
+If `gram` is not provided, for $n = 2k$ the form defined by $\begin{matrix} 0 & I_k \\ -I_k & 0 \end{matrix}$
+is used.
+It can be shown that for odd $n$ there is no non-degenerate skew-symmetric bilinear form on $R^n$.
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/LinearLieAlgebra.jl]:
```suggestion
For odd $n$ there is no non-degenerate skew-symmetric bilinear form on $R^n$.
```

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl [hunk 2]
```diff
@@ -317,37 +321,191 @@ function special_linear_lie_algebra(R::Field, n::Int)
...
+    for k in 1:dim
+      basis[k][i, 1:n] = ker[k, (i-1) * n .+ (1:n)]
+    end
+  end
+  return basis
+end
+
 @doc raw"""
-    special_orthogonal_lie_algebra(R::Field, n::Int) -> LinearLieAlgebra{elem_type(R)}
+    special_orthogonal_lie_algebra(R::Field, n::Int, gram::MatElem) -> LinearLieAlgebra{elem_type(R)}
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/LinearLieAlgebra.jl:346]:
```suggestion
    special_orthogonal_lie_algebra(R::Field, n::Int) -> LinearLieAlgebra{elem_type(R)}
    special_orthogonal_lie_algebra(R::Field, n::Int, gram::MatElem) -> LinearLieAlgebra{elem_type(R)}
```

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl [hunk 3]
```diff
@@ -317,37 +321,191 @@ function special_linear_lie_algebra(R::Field, n::Int)
...
+    form = zero_matrix(R, n, n)
+    form[1, 1] = one(R)
+    form[1 .+ (1:k), 1 + k .+ (1:k)] = identity_matrix(R, k)
+    form[1 + k .+ (1:k), 1 .+ (1:k)] = identity_matrix(R, k)
+  end
+  return special_orthogonal_lie_algebra(R, n, form)
+end
+
+@doc raw"""
+    symplectic_lie_algebra(R::Field, n::Int, gram::MatElem) -> LinearLieAlgebra{elem_type(R)}
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/LinearLieAlgebra.jl:439]:
```suggestion
    symplectic_lie_algebra(R::Field, n::Int) -> LinearLieAlgebra{elem_type(R)}
    symplectic_lie_algebra(R::Field, n::Int, gram::MatElem) -> LinearLieAlgebra{elem_type(R)}
```

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl [hunk 4]
```diff
@@ -317,37 +321,191 @@ function special_linear_lie_algebra(R::Field, n::Int)
...
+    special_orthogonal_lie_algebra(R::Field, n::Int, gram::Matrix) -> LinearLieAlgebra{elem_type(R)}
 
 Return the special orthogonal Lie algebra $\mathfrak{so}_n(R)$.
 
+Given a non-degenerate symmetric bilinear form $f$ via its Gram matrix `gram`,
+$\mathfrak{so}_n(R)$ is the Lie algebra of all $n \times n$ matrices $x$ over the field `R`
+such that $f(xv, w) = -f(v, xw)$ for all $v, w \in R^n$.
+
+If `gram` is not provided, for $n = 2k$ the form defined by $\begin{matrix} 0 & I_k \\ -I_k & 0 \end{matrix}$
+is used, and for $n = 2k + 1$ the form defined by $\begin{matrix} 1 & 0 & 0 \\ 0 & 0 I_k \\ 0 & I_k & 0 \end{matrix}$.
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/LinearLieAlgebra.jl:356]: Why that form in particular? Is that the form you used before?

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl [hunk 5]
```diff
@@ -317,37 +321,191 @@ function special_linear_lie_algebra(R::Field, n::Int)
...
+  form = map_entries(R, gram)
+  @req size(form) == (n, n) "Invalid matrix dimensions"
+  @req is_symmetric(form) "Bilinear form must be symmetric"
+  @req is_invertible(form) "Bilinear form must be non-degenerate"
+  basis = _lie_algebra_basis_from_form(R, n, form)
+  dim = length(basis)
+  @assert characteristic(R) != 0 || dim == div(n^2 - n, 2)
+  s = ["x_$(i)" for i in 1:dim]
+  L = lie_algebra(R, n, basis, s; check=false)
+  set_attribute!(L, :type => :special_orthogonal, :form => form)
```
Comment (fingolfin) [inline experimental/LieAlgebras/src/LinearLieAlgebra.jl:413]: Maybe also add a documented form to query the form? Of course that can also come in a future PR.

**Review by fingolfin — APPROVED** [review]

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl (hunk 5 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LinearLieAlgebra.jl:413]: Comes in a future PR

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl (hunk 4 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LinearLieAlgebra.jl:356]: No, I used the identity matrix before, which for general users is a bad choice, as eg over QQ the Cartan sub is not split.
This is the form used most often in the literature to construct the B_n and D_n types, eg in Humphreys Introduction to Lie Algebras. I think we should provide just any default here, and thus particular one has, in most cases, nice properties.

File: experimental/LieAlgebras/src/LinearLieAlgebra.jl (hunk 5 above)
Comment (lgoettgens) [reply experimental/LieAlgebras/src/LinearLieAlgebra.jl:413]: @fingolfin do you have a name idea for this function? `form` can be ambiguous with e.g. the killing form

Comment (thofma) [reply experimental/LieAlgebras/src/LinearLieAlgebra.jl:413]: `defining_form`?

Comment (thofma) [reply experimental/LieAlgebras/src/LinearLieAlgebra.jl:413]: `invariant_form` (if it is invariant :))?

**Pull request merged**